swat/cas/connection.py

```diff
@@ -434,51 +434,45 @@ def _detect_protocol(self, hostname, port, protocol=None):
         Returns
         -------
         string
-            'cas' or 'http'
+            'cas', 'http', or 'https'
 
         '''
+
         if protocol is None:
             protocol = cf.get_option('cas.protocol')
 
         if isinstance(hostname, six.string_types):
             hostname = re.split(r'\s+', hostname.strip())
 
         # Try to detect the proper protocol
+
         if protocol == 'auto':
 
             import socket
+            import ssl
 
-#           for ptype in ['http', 'https']:
             for host in hostname:
-                for ptype in ['http']:
-                    try:
-                        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-                        sock.settimeout(10)
-                        sock.connect((host, port))
-                        sock.send(('GET /cas HTTP/1.1\r\n' +
-                                   ('Host: %s\r\n' % host) +
-                                   'Connection: close\r\n' +
-                                   'User-Agent: Python-SWAT\r\n' +
-                                   'Cache-Control: no-cache\r\n\r\n').encode('utf8'))
-
-                        if sock.recv(4).decode('utf-8').lower() == 'http':
-                            protocol = ptype
-                            break
-
-                    except Exception as err:
-                        pass
-
-                    finally:
-                        sock.close()
-
-                    if protocol != 'auto':
-                        break
-
-                if protocol != 'auto':
-                    break
-
-            if protocol == 'auto':
-                protocol = 'cas'
+                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```

Comment:
There will need to be a break in here somewhere, otherwise it will always walk through all possible hostnames.  It should break out once it successfully finds a server that responds.  I'm not sure in this method how to tell between a server being there but the wrong protocol vs. the server not being there.

Reply:
@kesmit13, please see updated commit. I've completely rewritten the function and socket validations.